python/cudf/cudf/bindings/binops.pyx

```diff
@@ -165,29 +145,36 @@ def apply_op(lhs, rhs, out, op):
     return nullct
 
 
-def apply_op_udf(lhs, rhs, out, ptx):
+def apply_op_udf(lhs, rhs, udf_ptx, np_dtype):
```

Comment:
Can we make `np_dtype` an optional argument here? Unless the user specifies it, we should be able to compute a dtype for them based on the dtype of `lhs` and `rhs`.

Reply:
@shwina The output type has to be compatible with the operations inside the PTX code, otherwise we would have instances of memory leak. This is very different from having an operator name, say, `+`, and for `c = a + b`, `c` can have a variety of types. Also on the `libcudf` side the API is allocating the output column and returns it so we have to ask the user to provide the output type.

Reply:
> compute a dtype for them based on the dtype of lhs and rhs

@shwina the problem is that if you're doing a generic `op(lhs,rhs)`, how do you infer the type of the output? `op` can be anything.

That said, @hummingtree, can't you get the output type from the Numba generated device function? So ultimately, specifying output type can be hidden from the end-user because the output type can be determined by the Numba device function.

Reply:
@jrhemstad Yes numba device function does that and in the pytest I am having numba device function compiler to tell me the output type, which is then passed into the `apply_op_udf` without being specified explicitly.
Do you mean we want to have this in the python interface itself?

Reply:
> Do you mean we want to have this in the python interface itself?

Yes, like you said, `apply_op_udf` is not meant to be the final, user-facing interface. 

Pandas doesn't have an equivalent of doing a UDF binary op, so @kkraus14 has said it'll take some thinking to define the top-level interface. _That_ interface can take care of specifying the output data type to `apply_op_udf`.

Reply:
Sounds good. I guess we could just default to `float64`, but as you say, this is a discussion for another PR/issue :)